Support RSS streams (v4)

ejurgensen opened the pull request:
@whatdoineed2do here is a new edition of your RSS pr. As you can see, I have done some refactoring of especially rssscanner.c. You will see that I have dropped name and limit as parameters for adding items. I have trouble seeing the use of those. For limit, I have set a default of 10, but it is adjustable via pli->query_limit. I have also dropped item_remove, because I think once something has been added to the library it should be removed via the library ids.

Let me know if you find any mistakes. I haven't tested it a whole lot myself, but I will try to get that done. I would also like to ask if you think it might be better not to scan each item with scan_metadata_stream()? The first scan of those 10 items takes a while, and perhaps it would be ok to just add them based on the metadata in the RSS feed?

@chme could you review if the changes to the JSON api are ok? You are of course most welcome to also give other comments. I also observed the following, where you might have a solution: If I delete a playlist via the endpoint added here, it remains visible in the web UI. Only after Ctrl+F5 does it disappear.

whatdoineed2do commented on src/library/rssscanner.c:
```diff
@@ -0,0 +1,612 @@
+	}
+
+      scan_metadata_stream(&mfi, ri.url);
+
```
```
  if (mfi->song_length == 0 && mfi->file_size == 0)
    {
      DPRINTF(E_INFO, L_LIB, "Ignoring item (empty media) RSS id: %d name: '%s' url: %s pubdate: %s title: '%s'\n", pl_id, rss_feed_title, rss_item_url, rss_item_pubDate, rss_item_title);

      free_mfi(&mfi,1);
      continue;
    }
```
I would like to add this check back  - i've seen some elements that for some reason are published to the RSS stream but are actual empty media files.

whatdoineed2do replied:
Actually, this is one of the reasons why I went with scanning the underlying media instead of using the info from the RSS for title etc.

ejurgensen replied:
Can you share the RSS that has these empty files? I'm not sure it is forked-daapd's job to sanitize RSS's, but I suppose it depends on the reason they are empty. Also, having the check could lead to somewhat inconsistent behaviour if the file is unavailable during first time scan. Then it won't be added, but it will be added later if it comes online. Then it will stay there, even if it disappears again.

whatdoineed2do replied:
Trying to recall which RSS I saw this empty file but nothing there - may have been a temp issue with ffmpeg not reading from the remote server but I can't reproduce this right now.

However going through my RSS on this dev host, I see `http://cgpgrey.libsyn.com/rss` does have an RSS entry that has no `url` so we'll need
```
index 3a3e4827..d40e0c84 100644
--- a/src/library/rssscanner.c
+++ b/src/library/rssscanner.c
@@ -437,6 +437,12 @@ rss_save(struct playlist_info *pli, int *count, enum rss_scan_type scan_type)
          return -1;
        }
 
+      if (!ri.url)
+        {
+          DPRINTF(E_INFO, L_LIB, "Ignoring item (empty media) RSS id: %d name: '%s' url: %s pubdate: %s title: '%s'\n", pli->id, pli->title, ri.url, ri.pubdate, ri.title);
+          continue;
+        }
+
       db_pl_add_item_bypath(pli->id, ri.url);
       (*count)++;
```
The entry for this empty comes from `Thu, 24 Jan 2013` on this so you'll have to set `RSS_DEFAULT_LIMIT 100` to see this.

whatdoineed2do commented:
I had a test with this tree and it looks fine.  There's one comment about a check after `scan_metadata_stream()` that I think should be added but testing it with the podcast subscriptions I had in my db seemed to show no ill affects.

Will merge this into my main testing branch and let it sit on my RPI for a few days to see if anything comes up.

ejurgensen commented:
Good point, I've added the check for url as you suggested. Also added fixes for two other issues coming from my changes.

I notice that Valgrind says that the xml from mxmlLoadString is memleaking, which I don't understand since mxmlDelete() is being called as the docs say. I made a simple test program (see below), and Valgrind also says that is leaking. If it is true, the leak is quite large. So I need to find out what is going on.

```
int
main()
{
  FILE *fp;
  mxml_node_t *tree;
  mxml_node_t *node;

  fp = fopen("rss", "r");
  tree = mxmlLoadFile(NULL, fp, MXML_OPAQUE_CALLBACK);
  fclose(fp);

  if (!tree)
    {
      printf("Error\n");
      return -1;
    }

  node = mxmlFindElement(tree, tree, "item", NULL, NULL, MXML_DESCEND);
  if (!node)
    {
      printf("Didn't find item\n");
      return -1;
    }

  printf("Ok\n");

  mxmlDelete(tree);

  return 0;
}
```

whatdoineed2do commented:
The leak you see is a bit odd; what env?  I was testing this branch yday and I see no leaks via `valgrind` on this Fedora box -`valgrind 3.5.10` and `mxml-devel-3.0-1.fc30.x86_64`
```
$ gcc mx.c -lmxml -lpthread && valgrind --leak-check=full --show-leak-kinds=definite ./a.out 
==7965== Memcheck, a memory error detector
==7965== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==7965== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==7965== Command: ./a.out
==7965== 
Ok
==7965== 
==7965== HEAP SUMMARY:
==7965==     in use at exit: 0 bytes in 0 blocks
==7965==   total heap usage: 638 allocs, 638 frees, 39,409 bytes allocated
==7965== 
==7965== All heap blocks were freed -- no leaks are possible
==7965== 
==7965== For lists of detected and suppressed errors, rerun with: -s
==7965== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

$ cat mx.c 
#include <stdio.h>
#include <mxml.h>

int
main()
{
  FILE *fp;
  mxml_node_t *tree;

  fp = fopen("rss", "r");
  tree = mxmlLoadFile(NULL, fp, MXML_OPAQUE_CALLBACK);
  fclose(fp);

  if (!tree)
    {
      printf("Error\n");
      mxmlDelete(tree);
      return -1;
    }

  printf("Ok\n");

  mxmlDelete(tree);

  return 0;
}
```

ejurgensen commented:
Yes, very odd. It appears on Ubuntu 18.04 and on Raspbian Stretch, both with Valgrind 3.13.0 and mxml 2.10. Valgrind throws a lot of garbage on Raspbian Buster, so it is probably broken.

I suppose it is a false positive, but I will check the mxml 2.10 code to see what mxmlDelete does.

whatdoineed2do commented:
The same leak reported here https://github.com/michaelrsweet/mxml/issues/183 was fixed in 2.11 https://github.com/michaelrsweet/mxml/releases/tag/v2.11

ejurgensen commented:
Good find, too bad the memleak is real. It seems there is no CVE on this, and perhaps because of that no security update from Debian :-(

Since we are reading one or more xml's each hour, the memleak will quite large. So I'm thinking we either need to find a workaround for 2.10, or disable RSS for 2.10 and earlier. My preference is on the latter - also given that I don't know how to make a workaround.

whatdoineed2do commented:
Yes I think disabling RSS for <= 2.10 would be the best option.   Also looking at https://packages.debian.org/stretch/libmxml1 it seems to suggest that `stretch` will be at some point 3 releases old (bullseye, buster, stretch) so I guess this wouldn't hurt.  If anyone on `stretch` wanted this, they could build `mxml` themselves.

This small patch should be enough to validate:
```
diff --git a/configure.ac b/configure.ac
index 30e9d7c3..3ed09025 100644
--- a/configure.ac
+++ b/configure.ac
@@ -114,7 +114,12 @@ FORK_FUNC_REQUIRE([COMMON], [GNU libunistring], [LIBUNISTRING], [unistring],
 FORK_MODULES_CHECK([FORKED], [ZLIB], [zlib], [deflate], [zlib.h])
 FORK_MODULES_CHECK([FORKED], [CONFUSE], [libconfuse >= 3.0], [cfg_init], [confuse.h])
 FORK_MODULES_CHECK([FORKED], [MINIXML], [mxml], [mxmlNewElement], [mxml.h],
-       [AC_CHECK_FUNCS([mxmlGetOpaque] [mxmlGetText] [mxmlGetType] [mxmlGetFirstChild])])
+       [AC_CHECK_FUNCS([mxmlGetOpaque] [mxmlGetText] [mxmlGetType] [mxmlGetFirstChild])
+        dnl check for memleak fix in mxmlDelete
+        PKG_CHECK_EXISTS([mxml >= 2.11],
+               [AC_DEFINE([HAVE_MXML_NEW], 1,
+                       [Define to 1 if you have libmxml (>2.10)])], [])
+       ])
 
 dnl SQLite3 requires extra checks
 FORK_MODULES_CHECK([COMMON], [SQLITE3], [sqlite3 >= 3.5.0],
diff --git a/src/library.c b/src/library.c
index 98d37f7d..38325f2d 100644
--- a/src/library.c
+++ b/src/library.c
@@ -89,7 +89,9 @@ static struct library_source *sources[] = {
 #ifdef HAVE_SPOTIFY_H
     &spotifyscanner,
 #endif
+#ifdef HAVE_MXML_NEW
     &rssscanner,
+#endif
     NULL
 };
```

ejurgensen commented:
mxmlDelete() is used in multiple other parts of forked-daapd, so I changed my mind about just disabling, and instead made the above commit, where a fixed mxmlDelete() gets used (if mxml is <= 2.10). So that means RSS will also work for Raspbian Stretch + Ubuntu 18.04 users. It would be good if you could check that I haven't broken anything on your Fedora.

Other than that I am ready to merge this, but first I hope @chme can say if the json stuff is ok.

whatdoineed2do commented:
Ran this through valgrind on this fedora box - seems good.

whatdoineed2do commented on src/library/rssscanner.c:
```diff
@@ -0,0 +1,618 @@
+      free(mfi->title);
+      mfi->title = safe_strdup(ri->title);
+    }
+
```
I've been adding a few more podcasts and I've noticed that there are a few podcasts that publish mp3 but the mp3 od NOT incl any meta but the RSS stream has a valid title.  This means that the title of the just becomes the url.

Should we just add this:
```
  else                                                                                   
    {                                                                                    
      // Forcible take the RSS feed info if available                                    
      if (ri->title)                                                                     
        {                                                                                
          free(mfi->title);                                                              
          mfi->title = safe_strdup(ri->title);                                           
        }                                                                                
    }
```
For example, this RSS: `http://feeds.simplecast.com/61hlOb8v` feed has items:
```    <item>
      <guid isPermaLink="false">2798bc8d-8702-4b8b-aad4-bb54b4ac7c0c</guid>
      <title>Takeover Talk &amp; A Camp Nou Debut</title>
      <description><![CDATA[<p>1. Takeover Latest - 3:10</p><p>2. Should a PL club need govt help to pay staff wages? - 12:30</p><p>3. The Aaron Hughes Interview - 20:50</p><p>4. A Debut in The Camp Nou - 29:00</p><p>5. Aaron on Sir Bobby - 34:45</p><p>6. Souness & Leaving St James - 45:20</p>
]]></description>
      <pubDate>Wed, 1 Apr 2020 04:00:24 +0000</pubDate>
      <author>podcasts@theathletic.com (Chris Waugh, Aaron Hughes, Taylor Payne, George Caulkin)</author>
      <link>https://theathletic.com/podcast/147-pod-on-the-tyne/?episode=19</link>
      <content:encoded><![CDATA[<p>1. Takeover Latest - 3:10</p><p>2. Should a PL club need govt help to pay staff wages? - 12:30</p><p>3. The Aaron Hughes Interview - 20:50</p><p>4. A Debut in The Camp Nou - 29:00</p><p>5. Aaron on Sir Bobby - 34:45</p><p>6. Souness & Leaving St James - 45:20</p>
]]></content:encoded>
      <enclosure length="60243424" type="audio/mpeg" url="https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3?aid=rss_feed"/>
      <itunes:title>Takeover Talk &amp; A Camp Nou Debut</itunes:title>
      <itunes:author>Chris Waugh, Aaron Hughes, Taylor Payne, George Caulkin</itunes:author>
      <itunes:duration>01:02:36</itunes:duration>
      <itunes:summary>This week, Newcastle fan-favourite Aaron Hughes joins Taylor Payne, Chris Waugh &amp; George Caulkin on the pod, with George reporting that a long-anticipated club takeover is now at an advanced stage...

Aaron reflects on a fantastic career, which began in the Camp Nou against Barcelona; Playing for Sir Bobby Robson, Graeme Souness, while the guys relive the anger they felt when he was sold to Aston Villa!</itunes:summary>
      <itunes:subtitle>This week, Newcastle fan-favourite Aaron Hughes joins Taylor Payne, Chris Waugh &amp; George Caulkin on the pod, with George reporting that a long-anticipated club takeover is now at an advanced stage...

Aaron reflects on a fantastic career, which began in the Camp Nou against Barcelona; Playing for Sir Bobby Robson, Graeme Souness, while the guys relive the anger they felt when he was sold to Aston Villa!</itunes:subtitle>
      <itunes:keywords>sir bobby, barcelona, newcastle, premier league, football nostalgia, exclusive interview, coronavirus, aaron hughes, football, souness, podcast</itunes:keywords>
      <itunes:explicit>yes</itunes:explicit>
      <itunes:episodeType>full</itunes:episodeType>
      <itunes:episode>18</itunes:episode>
    </item>
```
and the underlying mp3
```
$ ffprobe -hide_banner -i https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3
Input #0, mp3, from 'https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3':
  Metadata:
    date            : 2020-03-31 13:25
    encoder         : Lavf58.35.100
    id3v2_priv.XMP  : <?xpacket begin="\xef\xbb\xbf" id="W5M0MpCehiHzreSzNTczkc9d"?>\x0a<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="Adobe XMP Core 5.6-c148 79.164050, 2019/10/01-18:03:16        ">\x0a <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">\x0a  <rdf
  Duration: 01:02:41.61, start: 0.025057, bitrate: 128 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, mono, fltp, 128 kb/s
```

ejurgensen replied:
Ok, I have made that change. I think the suggestion you have can be simplified to just always using the title from the RSS if there is one. Same rule for video, so the previous special rule for video isn't required any more.

ejurgensen commented:
I've merged this now, it would seem @chme doesn't have time to review right now.

Thanks for your work, @whatdoineed2do
